flux-desugar/src/table_resolver.rs

```diff
@@ -552,6 +599,10 @@ impl ResKey {
         let s = path.segments.iter().map(|segment| segment.ident).join("::");
         Ok(ResKey { s })
     }
+
+    fn from_lang_item(item: hir::LangItem) -> Self {
+        ResKey { s: format!("{:?}", item) }
```

Comment:
Let's not rely on the debug format
```suggestion
        ResKey { s: item.name().to_string() }
```

Reply:
Sadly had to use the `format` version here because the `to_string()` method returns 

```
TRACE: from_lang_item(Future) -> "future_trait"
```

which then fools the subsequent resolving (as we just write `Future<Item=...>`)

Reply:
@nilehmann check / advice ? ^^^

Reply:
@ranjitjhala 

Oh, I see. This is then related to https://github.com/flux-rs/flux/pull/480#discussion_r1297462221. I think the problem comes from supporting

```rust
#[flux::sig(fn() -> impl Future<Output=i32>)]
async fn foo() - i32 {
    0
}
```

which makes it weird because we have to resolve `Future` even though the user didn't write it. If we carry the asyncness down and generate the `impl Future<..>` during desugaring then we wouldn't have to resolve `Future`. During desugaring, we can create the`impl Future` by generating a `fhir::Res::Struct` with the appropriate `DefId`  corresponding to the `future_trait` lang item. If we do this then we won't need `fhir::Res::LangTrait`.

Reply:
quite so, let me make this change!

Reply:
Ok, take a look -- I didn't use the `fhir::Res::Struct` because post-desugaring, the `async` output is directly represented as an

1. `Opaque` (with the appropriate impl's `DefId`) 
2. the associated predicates `item_bounds` for (1)

So I modified the `resolve` to track (1) and generate (2) during `desugar_ty` (see `desugar_async_out_ty`)

Reply:
oh right, we have to resolve the `Opaque` type. This sounds good